### PR TITLE
component testing

### LINKED ISSUE #229: write component tests using Storybook + Cypress
https://storybook.js.org/docs/html/workflows/interaction-testing

### PR DESCRIPTION
**Pull request by dwhieb**
**Related Issue:**

closes #229

**Description of Changes**

Revamps the testing setup to using Storybook + Cypress in conjunction for writing component and integration tests. Unit tests still just use Mocha, while E2E tests just use Cypress.

Extensive documentation about the new testing setup has been added to #240.

### CODE REVIEW

**Review by vadekh — COMMENTED** [review]
@dwhieb It looks like this is all working, and each individual test command (`:component`, `:e2e`, etc.) works. However, the full `npm run test` command is very messy in console, because of the asynchronous nature of the command (I assume it is running all 4 tests at once). I wonder if there is any way to either force the commands to wait for previous call to finish (at some time cost), or if there is a way to silence some of the storybook info in console.

Two other issues I have noticed:
1) Sometimes `npm test` runs into an error where two of the commands are attempting to start the `6006` server, but if the server is started by the first command, the second command fails to start on the already occupied server.
2) When the error in (1) doesn't crash the test, the `npm test` command doesnt appear to terminate.

I think one possible solution is to break down the `npm test` command into smaller components so the test-type specific commands aren't overlapping. For example have the `npm test` command call the start server command once, and then call the `cypress-ct-run`/`cypress-it-run` commands instead of the full `test:component`/`test:integration` commands.

**Comment by dwhieb** [comment]
> @dwhieb It looks like this is all working, and each individual test command (`:component`, `:e2e`, etc.) works.

🎉!

> However, the full `npm run test` command is very messy in console, because of the asynchronous nature of the command (I assume it is running all 4 tests at once). I wonder if there is any way to either force the commands to wait for previous call to finish (at some time cost), or if there is a way to silence some of the storybook info in console.

I _think_ changing `&` to `&&` in the npm script should fix this. It forces commands to run in sequence rather than in parallel. (Apparently on Windows there's no real difference?) Can you try this and see if it improves the output?

If the above change works, I think it'll fix the other two issues you listed too.

> I think one possible solution is to break down the `npm test` command into smaller components so the test-type specific commands aren't overlapping. For example have the `npm test` command call the start server command once, and then call the `cypress-ct-run`/`cypress-it-run` commands instead of the full `test:component`/`test:integration` commands.

Definitely. Even if the above fix works, this would still be good to do. It definitely bothers me that Storybook gets started multiple times.

I think if the above fix works, I'll update the `npm test` script as part of this PR, rather than opening a separate issue for it.

**Comment by vadekh** [comment]
Yup, the `&&` change fixed the errors. Storybook still starts twice of course, which also prints a lot of meaningless lines to console in between the tests, so for readability we should try to start/run those tests first. Then the test result outputs should all be printed more or less in a row.

File: package.json [hunk 1]
```diff
@@ -19,24 +19,31 @@
...
+    "cypress-e2e-open": "cypress open --config-file test/cypress.e2e.json",
+    "cypress-e2e-run" : "cypress run --config-file test/cypress.e2e.json",
+    "cypress-it"      : "start-server-and-test start:storybook 6006 cypress-it-open",
+    "cypress-it-open" : "cypress open --config-file test/cypress.integration.json",
+    "cypress-it-run"  : "cypress run --config-file test/cypress.integration.json",
+    "start"           : "node build/bin/startServer.js",
+    "start:storybook" : "start-storybook -p 6006 -s dist --ci",
+    "stop-only"       : "stop-only -f test,src",
+    "storybook"       : "start-storybook -p 6006",
+    "test"            : "npm run test:unit & npm run test:component & npm run test:integration & npm run test:e2e",
```
Comment (vadekh) [inline package.json]:
```suggestion
    "test"            : "npm run test:unit && npm run test:component && npm run test:integration && npm run test:e2e",
```

**Review by vadekh — CHANGES_REQUESTED** [review]

**Review by vadekh — APPROVED** [review]

**Pull request merged**